packages/xforms-engine/src/instance/RankControl.ts

```diff
@@ -0,0 +1,144 @@
+
+type AssertRangeNodeDefinition = (definition: RankDefinition) => asserts definition is RankDefinition<'string'>;
+const assertRangeNodeDefinition: AssertRangeNodeDefinition = (definition) => {
+	if (definition.valueType !== 'string') {
+		throw new RankValueTypeError(definition);
+	}
+};
+
+type RankItemMap = ReadonlyMap<string, RankItem>;
+
+interface RankControlStateSpec extends ValueNodeStateSpec<readonly string[]> {
+	readonly label: Accessor<TextRange<'label'> | null>;
+	readonly hint: Accessor<TextRange<'hint'> | null>;
+	readonly valueOptions: Accessor<RankValueOptions>;
+}
+
+export class RankControl
+	extends ValueNode<'string', RankDefinition<'string'>, readonly string[], readonly string[]>
+	implements
+		RankNode,
+		XFormsXPathElement,
+		EvaluationContext,
+		ValidationContext,
+		ClientReactiveSubmittableValueNode
+{
+	static from(parent: GeneralParentNode, definition: RankDefinition): RankControl {
+		assertRangeNodeDefinition(definition);
+		return new this(parent, definition);
+	}
+
+	private readonly mapOptionsByValue: Accessor<RankItemMap>;
+
+	protected override readonly getInstanceValue: Accessor<string>;
+
+	// XFormsXPathElement
+	override readonly [XPathNodeKindKey] = 'element';
+
+	// InstanceNode
+	protected readonly state: SharedNodeState<RankControlStateSpec>;
+	protected readonly engineState: EngineState<RankControlStateSpec>;
+
+	// RankNode
+	readonly nodeType = 'rank';
+	readonly currentState: CurrentState<RankControlStateSpec>;
+
+	private constructor(parent: GeneralParentNode, definition: RankDefinition<'string'>) {
+		const codec = new BaseItemCollectionCodec(sharedValueCodecs.string);
+		super(parent, definition, codec);
+
+		const valueOptions = createItemCollection(this);
+		const mapOptionsByValue: Accessor<RankItemMap> = this.scope.runTask(() => {
+			return createMemo(() => {
+				return new Map(valueOptions().map((item) => [item.value, item]));
+			});
+		});
+
+		this.mapOptionsByValue = mapOptionsByValue;
+
+		const baseValueState = this.valueState;
+		const [baseGetValue, setValue] = baseValueState;
+		const getValue = this.scope.runTask(() => {
+			return createMemo(() => {
+				const optionsByValue = mapOptionsByValue();
+				return baseGetValue().filter((value) => optionsByValue.has(value));
+			});
+		});
+		const valueState: SimpleAtomicState<readonly string[]> = [getValue, setValue];
+
+		this.getInstanceValue = this.scope.runTask(() => {
+			return createMemo(() => codec.encodeValue(getValue()));
+		});
+
+		const sharedStateOptions = {
+			clientStateFactory: this.engineConfig.stateFactory,
+		};
+
+		const state = createSharedNodeState(
+			this.scope,
+			{
+				reference: this.contextReference,
+				readonly: this.isReadonly,
+				relevant: this.isRelevant,
+				required: this.isRequired,
+				label: createNodeLabel(this, definition as AnyNodeDefinition),
+				hint: createFieldHint(this, definition as AnyNodeDefinition),
```

Comment:
```suggestion
				label: createNodeLabel(this, definition),
				hint: createFieldHint(this, definition),
```

Only one of these is a type error, but both are unnecessary and overly broad.

Sometimes it really is necessary or just the least-bad option to type cast, but oftentimes it masks other problems. It's usually worth exploring other ways to work with the type system before overriding it.

(If I had to guess, these casts were probably added to work around other type errors, possibly around the mismatched `ValueType` generic?)

Reply:
Agree! Fixed